internal/cli/cmd/apply/apply.go

```diff
@@ -19,8 +19,54 @@ import (
 	kargosvcapi "github.com/akuity/kargo/pkg/api/service/v1alpha1"
 )
 
+type applyOptions struct {
+	*option.Option
+
+	Filenames []string
+}
+
+// newApplyOptions creates a new options struct.
+func newApplyOptions(opt *option.Option) *applyOptions {
+	return &applyOptions{Option: opt}
+}
+
+// AddFlags adds the flags for the apply options to the provided command.
+func (o *applyOptions) AddFlags(cmd *cobra.Command) {
+	o.PrintFlags.AddFlags(cmd)
+
+	// TODO: Factor out server flags to a higher level (root?) as they are
+	//   common to almost all commands.
+	option.InsecureTLS(cmd.PersistentFlags(), o.Option)
+	option.LocalServer(cmd.PersistentFlags(), o.Option)
+
+	option.Filenames(cmd.Flags(), &o.Filenames, "Filename or directory to use to apply the resource(s)")
+
+	if err := cmd.MarkFlagRequired(option.FilenameFlag); err != nil {
+		panic(errors.Wrap(err, "could not mark filename flag as required"))
+	}
+	if err := cmd.MarkFlagFilename(option.FilenameFlag, ".yaml", ".yml"); err != nil {
+		panic(errors.Wrap(err, "could not mark filename flag as filename"))
+	}
+	if err := cmd.MarkFlagDirname(option.FilenameFlag); err != nil {
+		panic(errors.Wrap(err, "could not mark filename flag as dirname"))
+	}
+}
+
+// Validate performs validation of the options. If the options are invalid, an
+// error is returned.
+func (o *applyOptions) Validate() error {
```

Comment:
Kudos for this...

I've frequently seen this pattern of using a struct to encapsulate options and have rarely seen much value in it...

Coupled with the decision to extract validation into its own function, that pattern begins to make sense.

You've converted me.

Nit: Could be unexported.